homeassistant/components/fan/template.py

```diff
@@ -0,0 +1,339 @@
+
+    @property
+    def speed_list(self: ToggleEntity) -> list:
+        """Get the list of available speeds."""
+        return self._speed_list
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self._state
+
+    @property
+    def speed(self):
+        """Return the current speed."""
+        return self._speed
+
+    @property
+    def oscillating(self):
+        """Return the oscillation state."""
+        return self._oscillating
+
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return False
+
+    # pylint: disable=arguments-differ
+    async def async_turn_on(self, speed: str = None) -> None:
+        """Turn on the fan.
+
+        This method is a coroutine.
+        """
+        self._state = True
+        await self._on_script.async_run()
+
+        if speed:
+            await self.async_set_speed(speed)
+
+    # pylint: disable=arguments-differ
+    async def async_turn_off(self) -> None:
+        """Turn off the fan.
+
+        This method is a coroutine.
+        """
+        self._state = False
+        await self._off_script.async_run()
+
+    async def async_set_speed(self, speed: str) -> None:
+        """Set the speed of the fan.
+
+        This method is a coroutine.
+        """
+        if self._set_speed_script is None:
+            return
+
+        if speed in self._speed_list:
+            self._speed = speed
+            await self._set_speed_script.async_run({ATTR_SPEED: speed})
+        else:
+            _LOGGER.error(
+                'Received invalid speed: %s. ' +
+                'Expected: %s.',
+                speed, self._speed_list)
+
+    async def async_oscillate(self, oscillating: bool) -> None:
+        """Set oscillation of the fan.
+
+        This method is a coroutine.
+        """
+        if self._set_oscillating_script is None:
+            return
+
+        if oscillating is True or oscillating is False:
+            self._oscillating = oscillating
+            await self._set_oscillating_script.async_run(
+                {ATTR_OSCILLATING: oscillating}
+            )
+        else:
+            _LOGGER.error(
+                'Received invalid oscillating: %s. ' +
+                'Expected True/False.', oscillating)
+
+    async def async_added_to_hass(self):
+        """Register callbacks."""
+        @callback
+        def template_fan_state_listener(entity, old_state, new_state):
+            """Handle target device state changes."""
+            self.async_schedule_update_ha_state(True)
+
+        @callback
+        def template_fan_startup(event):
+            """Update template on startup."""
+            self.hass.helpers.event.async_track_state_change(
+                self._entities, template_fan_state_listener)
+
+            self.async_schedule_update_ha_state(True)
+
+        self.hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_START, template_fan_startup)
```

Comment:
I don't understand why we should only call this on `EVENT_HOMEASSISTANT_START`. This code would break once we start dynamically loading/unloading entities. Maybe have a look at how it's done in the light group platform:

https://github.com/home-assistant/home-assistant/blob/ca5f4709564773c8cebd6ee40aa7cc1095b19105/homeassistant/components/light/group.py#L72-L82

Reply:
I am doing the same as in `light/template` and `sensor/template`. Using the code in light group did not work. Fan is no changed when other entities changed